feat: add mTLS support for model providers

asm89 opened the pull request:
Adds mutual TLS (mTLS) authentication support for model provider connections, enabling Codex to connect to endpoints requiring client certificates.

- Ensure rustls-only TLS backend across the workspace
- Add `tls` config block for CA and client certificates
- Support reading cert paths from environment variables
- Add `just check-tls` to verify no native-tls in dependency tree

asm89 commented:
I have read the CLA Document and I hereby sign the CLA

asm89 commented on codex-rs/core/src/default_client.rs:
```diff
@@ -262,22 +275,137 @@ pub fn create_client() -> CodexHttpClient {
+            Err(e) => {
+                tracing::error!("Failed to apply TLS configuration: {}", e);
+                // Fall back to base builder without TLS
+                create_base_client_builder()
```
As noted in the commit description, the existing code already had a fallback on builder failures.

That said, agreed. Making `create_default_client()` fallible is a bigger change across the codebase however. So hopeful these fixes can be shipped separately?

Perhaps failing harder is okay for the time being?

bolinfest reviewed (commented):
Can you please also update https://github.com/openai/codex/blob/main/docs/config.md as part of this PR?

asm89 commented:
@bolinfest How about I remove the relative path loading from this PR and only support absolute paths? I'm a bit worried about getting stuck rebasing this for a while. FWIW for our enterprise deployment I really need commit 1 (making mTLS actually work!) and env var based loading.

As for the path resolving as implemented in this PR. I tried to mirror what OTEL does, but you're right that it incorrectly always resolves against codex home at the moment. Not something relative to the config file. To properly support the latter, I think something closer to my comment above would be needed.

bolinfest commented:
@asm89 Admittedly, getting the config stuff correct is tricky in its current form, but https://github.com/openai/codex/pull/7796 should fix things?

asm89 committed:
fix: ensure rustls backend is used for mTLS support

The mTLS implementation was failing with "tlsv13 alert certificate
required" errors despite certificates loading correctly. The root cause
was Cargo's feature unification - when multiple crates depend on
reqwest, Cargo enables the union of all requested features. Since some
crates were using default features (which includes native-tls) and
others specified rustls, both TLS backends were being compiled and
linked, causing unpredictable behavior during TLS handshakes.

The fix ensures rustls-only is used across the entire workspace by
setting `default-features = false` and explicitly specifying `features =
["rustls-tls"]` for all reqwest dependencies. This is important because
the mTLS code uses `Identity::from_pem()` which only exists with rustls.

Added a verification script (`just check-tls`) that fails if openssl-sys
appears in the dependency tree.

asm89 committed:
feat: add mTLS support for model providers

Adds mutual TLS (mTLS) authentication support for model provider
connections, allowing Codex to connect to endpoints requiring
client certificates. The implementation uses rustls exclusively -
native-tls is not supported due to Cargo's feature unification causing
both TLS backends to be linked when any dependency enables native-tls.

Configuration is done via TOML for each model provider:
```toml
[model-providers.my-provider]
base_url = "https://example.com"
tls.ca-certificate = "/path/to/ca.pem"
tls.client-certificate = "/path/to/client.pem"
tls.client-private-key = "/path/to/key.pem"
```

Relative paths are resolved against the config file's directory during
config loading via AbsolutePathBuf. All three TLS fields are optional -
omit CA certificate to use system roots, omit client cert/key for
server-only TLS.

The implementation gracefully falls back to a non-TLS client if certificate
loading fails (with error logging), ensuring connectivity isn't completely
broken by misconfigured certificates. This matches the previous behavior on
reqwest builder failures.

asm89 committed:
docs: document TLS/mTLS configuration for model providers

asm89 committed:
feat: support reading TLS cert paths from env vars

Adds support for specifying TLS certificate paths via environment
variables in model provider configuration. Each certificate/key path
field now has a corresponding `-env` variant that names an environment
variable containing the path.

New config fields:
- `ca-certificate-env`: env var for CA certificate path
- `client-certificate-env`: env var for client certificate path
- `client-private-key-env`: env var for client private key path

If both a direct path and env var are specified, the env var takes
precedence (if set and non-empty). This follows the same pattern used
by `env_http_headers` for HTTP headers in the codebase.

Example config:
```toml
[model_providers.my-provider.tls]
ca-certificate-env = "MY_CA_CERT_PATH"
client-certificate-env = "MY_CLIENT_CERT_PATH"
client-private-key-env = "MY_CLIENT_KEY_PATH"
```

The value of the environment variable has to be an absolute path.

bolinfest commented on codex-rs/core/src/default_client.rs:
```diff
@@ -134,22 +135,134 @@ pub fn create_client() -> CodexHttpClient {
+}
+
+pub fn build_configured_reqwest_client(tls_config: Option<&TlsConfig>) -> reqwest::Client {
+    let builder = create_base_client_builder();
```
Why do this and then immediately reassign?

bolinfest commented on codex-rs/core/src/default_client.rs:
```diff
@@ -134,22 +135,134 @@ pub fn create_client() -> CodexHttpClient {
+    let builder = create_base_client_builder();
+
+    // Apply TLS configuration if provided
+    let builder = if let Some(tls) = tls_config {
```
I would favor `match` and in the `None` case do `create_base_client_builder()`

bolinfest commented on codex-rs/core/src/default_client.rs:
```diff
@@ -134,22 +135,134 @@ pub fn create_client() -> CodexHttpClient {
+        match apply_tls_config(builder, tls) {
+            Ok(configured_builder) => configured_builder,
+            Err(e) => {
+                tracing::error!("Failed to apply TLS configuration: {}", e);
```
```suggestion
                tracing::error!("Failed to apply TLS configuration: {e}");
```

bolinfest commented on codex-rs/core/src/default_client.rs:
```diff
@@ -134,22 +135,134 @@ pub fn create_client() -> CodexHttpClient {
+                "Failed to read CA certificate from {}: {}",
+                ca_path.display(),
+                e
+            )
```
I feel like our clippy rules should catch this?

```suggestion
            format!(
                "Failed to read CA certificate from {}: {e}",
                ca_path.display(),
            )
```

bolinfest commented on codex-rs/core/src/model_provider_tls_config.rs:
```diff
@@ -0,0 +1,214 @@
+    }
+
+    // Fall back to direct path (already absolute from deserialization)
+    direct.cloned()
```
```suggestion
    {
        Some(path)
    } else {
        // Fall back to direct path (already absolute from deserialization)
        direct.cloned()
    }
```

bolinfest commented on codex-rs/core/src/model_provider_info.rs:
```diff
@@ -96,9 +98,20 @@ pub struct ModelProviderInfo {
+    /// Build a reqwest client configured for this model provider.
+    /// This extracts TLS configuration from the provider if present.
+    pub fn build_reqwest_client(&self) -> reqwest::Client {
+        let tls_config = self.tls.as_ref().map(ModelProviderTlsConfig::to_tls_config);
```
Something like:

```suggestion
        let tls_config = self.tls.map(|m| m.clone().into());
```
